ユーザーフォロー機能を作成しました。

naomichi-h opened the pull request:
Ruby on Railsのプラクティスサンプルアプリに、ユーザーフォロー機能を追加しました。
rubocop通過ずみです。

レビューのほど、よろしくお願いいたします。

naomichi-h committed:
add:ユーザーフォロー機能を追加しました。
relationshipsテーブルのための、新たなマイグレーションファイルを追加しました。
relationshipモデルおよびrelationshipsコントローラを追加しました。
userモデルおよびusersコントローラにメソッドとアクションを追加しました。
routes.rbに新たなルーティングを記述しました。
show.html.erbにフォローボタンとフォロー数、フォロワー数の表示リンクを追加しました。
フォロー一覧ページのためのfollowing.html.erbを追加しました。
フォロワー一覧ページのためのfollower.html.erbを追加しました。

naomichi-h committed:
fix:rubocopに通しました。
マイグレーションファイルで一部not reversibleになっていた記述を修正しました。

remove:不要なファイルを削除しました。
relationshipに関する不要なtestおよびscssファイルを削除しました。

update:フォロー一覧とフォロワー一覧ページに表題を追加しました。

yoshitsugu commented on app/controllers/relationships_controller.rb:
```diff
@@ -0,0 +1,14 @@
+
+class RelationshipsController < ApplicationController
+  def create
+    user = User.find_by(id: params[:user_id])
```
`find_by` だと指定のidのレコードがなかった場合にnilになり、後続の処理がエラーになってしまうので、 `find` を使って、 `ActiveRecord::NotFound` を発生させる(本番では404になるようにする)か、nilだった場合に404表示を返すような処理を追加するのがよく見るパターンですかね〜

yoshitsugu commented on app/controllers/relationships_controller.rb:
```diff
@@ -0,0 +1,14 @@
+  def create
+    user = User.find_by(id: params[:user_id])
+    current_user.follow(params[:user_id]) unless current_user.following?(user)
+    redirect_to root_path
```
フォロー/アンフォローの状態変更だけなので、そのユーザーのページにリダイレクトしたほうが直感的かなと思います

yoshitsugu commented on app/controllers/relationships_controller.rb:
```diff
@@ -0,0 +1,14 @@
+
+  def destroy
+    Relationship.find(params[:id]).destroy
+    redirect_to root_path
```
こちらもユーザー詳細にリダイレクトしましょうか

yoshitsugu commented on db/migrate/20210712064516_create_relationships.rb:
```diff
@@ -0,0 +1,14 @@
+class CreateRelationships < ActiveRecord::Migration[6.1]
+  def change
+    create_table :relationships do |t|
+      t.references :following_user, foreign_key: { to_table: :users }, null: false
+      t.references :followed_user, foreign_key: { to_table: :users }, null: false
+
```
foreign keyも使えていてナイスです :+1:

yoshitsugu commented on db/migrate/20210712064516_create_relationships.rb:
```diff
@@ -0,0 +1,14 @@
+
+      t.timestamps
+
+      t.index %i[following_user_id followed_user_id], unique: true
```
:+1:

yoshitsugu commented on app/models/relationship.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class Relationship < ApplicationRecord
```
Relationshipだとちょっと意味として汎用的すぎるかもしれません。 `Following` とかでどうでしょうか

yoshitsugu commented on app/controllers/users_controller.rb:
```diff
@@ -7,5 +7,16 @@ def index
+
+  def followings
+    @user = User.find(params[:id])
+    @followings = @user.following_users
```
~細かいですが、アバターのN+1問題が発生していそうですので確認おねがいします！~ まちがってました！

yoshitsugu commented on app/controllers/users_controller.rb:
```diff
@@ -7,5 +7,16 @@ def index
+
+  def followers
+    @user = User.find(params[:id])
+    @followers = @user.followed_users
```
~こちらもアバターのN+1問題が発生していそうです！~ まちがってました！

yoshitsugu reviewed (commented):
気になったところコメントしました！

naomichi-h committed:
fix:レビューを受けて以下の変更を行いました

find_byではなく、findを使用するように変更しました。
フォロー/アンフォロー後のリダイレクト先をユーザーのページに変更しました。
中間テーブル名をrelationshipsからfollowingsに変更しました。

naomichi-h commented:
レビューを受けて、修正を行いました！
確認のほど、よろしくお願いいたします。

テーブル名の変更については、以下のサイトを参考に行いました。
> [【Ruby on Rails】テーブル名の変更方法と変更時にやること一覧＆ステップ \- Qiita](https://qiita.com/mtoyopet/items/a8558c9b7010dde6ecf8)

yoshitsugu commented on app/controllers/followings_controller.rb:
```diff
@@ -0,0 +1,14 @@
+
+  def destroy
+    Following.find(params[:id]).destroy
+    redirect_to controller: :users, action: :show, id: params[:id]
```
こちらは `Following` のidなので、リダイレクト先が関係ないユーザーの詳細ページになってしまっているようです！

yoshitsugu reviewed (commented):
動作確認していて気になったところコメントしました〜

naomichi-h committed:
fix:フォロー解除した際のリダイレクト先が、無関係なユーザーの詳細ページになっていた問題を修正しました。

naomichi-h commented:
フォローを解除した際のリダイレクト先が関係ないユーザーの詳細ページになっていた問題を修正しました！
確認のほど、よろしくお願いいたします。

yoshitsugu reviewed (approved):
確認しました！OKです！
